OPEN DATA SCIENCE
## PORTFOLIO 1 Titanic
Gress-Wright tsp932
5 January 2020


### INTRODUCTION

This Portfolio I consists of this introduction, an analysis, Python computer code and its output.

The task of Portfolio I involves analyzing data on passengers on the Titanic. There are several steps involved, including transforming, checking and teasing out the necessary data, describing the resulting data file, followed by the creation of descriptive statistics and then data analysis.  The Python computer code was originally produced in the development environment Spyder. The final version here is in Jupyter Notebook.

In the course overall, we have learned to work with large quantities of data using Python, a high level programming language. We have used it to generate descriptive statistics and for text mining.  Python and its methods are a toolkit, a means to a research end (Downey, 2015; Ignatow and Milhacea, 2018). They help the practice of Open Data Science. This Introduction therefore sets the portfolio as a whole in the context of Open Data Science (Wilkinson, et al., 2016).


ABOUT OPEN DATA SCIENCE: BACKGROUND, KNOWLEDGE ORGANIZATION AND ETHICS

Open Science is defined as “transparent and accessible knowledge that is shared and developed through collaborative networks” (Vicente-Saez & Martinez-Fuentes 2018, abstract).

Open Data Science is the application of Open Data Science (ODS) principles to data science and the entire research life cycle (from idea gathering to final publication) from day one. Thus ODS facilitates both on-line research collaboration using open-source software and large, digitally accessible data sets but also replication and dissemination of results without unnecessary duplication (Lund, 2019).  

Technological and institutional forces both drive the transition to Open Data Science but also hinder it. Data are increasingly to be found, and required to be available, in "computer artifacts” e.g. online lab notebooks, code, scripts, datasets, algorithms and software (Association for Computing Machinery, 2018). Funders increasingly demand, not a final product such as a project report or a published article, but access to the underlying data itself. On the other hand, the publish or perish culture of universities favors keeping one's data to oneself.  Technology makes security and privacy difficult to ensure and commercial pressures may favor unethical practices.


FAIR  is the attempt to “improve the infrastructure supporting the reuse of scholarly data” (Wilkinson et al., 2016: 1). FAIR is designed with digital technology in mind.   “Distinct from peer initiatives that focus on the human scholar, the FAIR Principles put specific emphasis on enhancing the ability of machines to automatically find and use the data, in addition to supporting its reuse by individuals.” (Wilkinson et al., 2016: 1).  FAIR is applied to the research cycle and helps to determine what practices and tools to use at each stage (Lund, 2019). Infrastructure should ensure that data is:

    Findable
    Accessible
    Interoperable (important!)
    Refindable


ETHICS IN OPEN DATA SCIENCE

As well, there are ethical considerations in Open Science and Open Data Science. These are expressed in the ACM Code of Ethics and Professional Conduct.  The Code as a whole is "concerned with how fundamental ethical principles apply to a computing professional's conduct... [and]...serves as a basis for ethical decision-making.” (Anderson, 1992).  The six core principles are:

1.1 Contribute to society and to human well-being, acknowledging that all people are stakeholders in computing.
1.2 Avoid harm.
1.3 Be honest and trustworthy.
1.4 Be fair and take action not to discriminate.
1.5 Respect the work required to produce new ideas, inventions, creative works, and computing artifacts.
1.6 Respect privacy.

In practice, these include, for example, the duty to secure informed consent and to ensure privacy (Lund, 2019). 

Yet, in a digital world,  following these principles is not always easy when are commercial, political or even criminal pressures involved.  For example,  Principle 1.6  requires that sensitive personal information not be traceable to the individuals involved.  Data should be anonymized.  Yet a New York Times investigation of a single data leak (50 billion pings of app-derived location data) revealed that the data, contrary to formal policy, was neither anonymous nor secure (Thompson and Warzel, 2019). These pressures mean that a personal commitment to ethical conduct is more important than ever.


### ASSIGNMENT

I open the data file "titanic.csv" in a texteditor using Windows Notepad and look at the contents. This allows me to visually inspect all the data and datatypes. There are integers, floats and strings.
The actual data seem complete for the 887 persons. BUT there were 2224 passengers and crew so data for about three fifths of the passengers is missing.  


Next, I import pandas. Next, I create the variable "df" for the dataframe. I do not specify a separator because this is a .csv file and pandas assume comma separators by default.


In [71]:
print('STEP 1:  Import pandas') 
import pandas as pd
df = pd.read_csv("titanic.csv") 
print(df)

STEP 1:  Import pandas
     Survived  Pclass                                               Name  \
0           0       3                             Mr. Owen Harris Braund   
1           1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2           1       3                              Miss. Laina Heikkinen   
3           1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4           0       3                            Mr. William Henry Allen   
..        ...     ...                                                ...   
882         0       2                               Rev. Juozas Montvila   
883         1       1                        Miss. Margaret Edith Graham   
884         0       3                     Miss. Catherine Helen Johnston   
885         1       1                               Mr. Karl Howell Behr   
886         0       3                                 Mr. Patrick Dooley   

        Sex   Age  Siblings/Spouses Aboard  Parents/Children Abo

2a. Data types

i. Datatype of data frame itself 
To find the variable type for df itself, I use the 'type' function. 'df' is a <class 'pandas.core.frame.DataFrame'>

ii. Data types of columns 
Next, I used print(df.dtypes) and (df.info()), which gives data type by column, and sums up the total of each data type. It showed integers, floats, and strings (e.g. Survived cell datatype was 0 or 1 i.e. integer, while Name cell data type was a string - identified as an 'object')


In [72]:
print('Data types')
print(type(df))
print(df.dtypes)
print(df.info()) 

Data types
<class 'pandas.core.frame.DataFrame'>
Survived                     int64
Pclass                       int64
Name                        object
Sex                         object
Age                        float64
Siblings/Spouses Aboard      int64
Parents/Children Aboard      int64
Fare                       float64
dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887 entries, 0 to 886
Data columns (total 8 columns):
Survived                   887 non-null int64
Pclass                     887 non-null int64
Name                       887 non-null object
Sex                        887 non-null object
Age                        887 non-null float64
Siblings/Spouses Aboard    887 non-null int64
Parents/Children Aboard    887 non-null int64
Fare                       887 non-null float64
dtypes: float64(2), int64(4), object(2)
memory usage: 55.6+ KB
None


2b. Shape of the dataframe(numbers of rows and colums) First I ask how many rows, then how many rows and columns (returned as a tuple) and how many cells. I then ask what are the columns' names.


In [73]:
print('Describe the dataframe') 
print('Number of rows: ', len(df))
print('Number of rows and columns: ', df.shape)
print('Number of cells: ', df.size)
print('Column names: ', df.columns)

Describe the dataframe
Number of rows:  887
Number of rows and columns:  (887, 8)
Number of cells:  7096
Column names:  Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'Siblings/Spouses Aboard',
       'Parents/Children Aboard', 'Fare'],
      dtype='object')


3)  

3.1 I first extract chosen data into a new file. I then calculate, using thosedata, the descriptive statistics for the 887 passengers. I chose to do this by creating a pivot table from dataframe df.


3a. The number of surviving passengers


In [74]:
print('Describe the passengers')
print('The number of surviving passengers')

pivot1=pd.pivot_table(df, index=['Survived'], values=['Name'], aggfunc={'count'})
print('Surviving passengers (0 = dead, 1 = survived)')
print(pivot1)
print('_____________________________________________________________________') 


Describe the passengers
The number of surviving passengers
Surviving passengers (0 = dead, 1 = survived)
          Name
         count
Survived      
0          545
1          342
_____________________________________________________________________


The pivot_table function indexed (i.e. sorted) the lines (i.e. records) by the values in column "Survived" (i.e. 0 died, 1 survived). This gave two groups. Values is taken from column "Name", in order to count something which was not a number. Aggfunc={count} sums up the total number of names in each group 0 = 545, 1 = 342. The number of surviving passengers is 342.


3b. The mean age for ALL passengers


In [75]:
mean = df.mean(axis=0)
print('Mean age, for ALL passengers')
print(mean['Age'])
print('_____________________________________________________________________')

Mean age, for ALL passengers
29.471443066516347
_____________________________________________________________________


3c. The median age for ALL passengers


In [76]:
median = df.median(axis=0)
print('Median age, for ALL passengers')
print(median['Age'])
print('_____________________________________________________________________')

Median age, for ALL passengers
28.0
_____________________________________________________________________


3d.  EXTRA The mean and median age for passengers, by survival status
Did age correlate with survival?


The median age, 28.0. is the same for both survivors and dead. However, the mean age is slightly higher for those that died. In light of the median, that means that the dead included more that were 'especially' old, compared to the survivor group.


In [77]:
print('Mean and median age of passengers')

pivot2=pd.pivot_table(df, index=['Survived'], values=['Age'], aggfunc={'mean','median'})
print('Mean and median age of passengers, by survival status(0 = died,1 = survived)')
print(pivot2)
print('_____________________________________________________________________') 


Mean and median age of passengers
Mean and median age of passengers, by survival status(0 = died,1 = survived)
                Age       
               mean median
Survived                  
0         30.138532   28.0
1         28.408392   28.0
_____________________________________________________________________


4)
The purpose here is to find out whether there are multiple instances of a given last name.
First,  identify each unique last name. For example, “Carr“ must not be confused with “Carrou”. Second, find a way to identify multiple instances of a given name. In order to do that, one must check, for each unique last name, whether there are any matches.


The challenge in this data is to identify the unique last names. In the .csv file, last names are not distinguished from first and middle names. Instead, each passengers’ full name is listed in a single string e.g. 'Mr. John Henry Rutherford'. The married women are registered with both their own first and middle names and their husband’s e.g. 'Mrs. Charles (Catherine Anne) Smith' 

To identify what is a last name, I read entries for 'Name' into a separate list of all the names. Then for each entry I split the name with 'rsplit'. For each line, I get the last name, split at the last whitespace into two segments they are stored into ‘last_name’. Each second segment is stored in ‘ls’. In order to avoid confusing names like “Carr” and “Carrou”, I add a space to the 'ls'.


To distinguish between unique and multiple instances,  I first create an empty list “first_time_seen”. Ultimately first_time_seen is going to contain one and only one instance of each and every last name.  Second, I identify multiple instances by creating a second empty list “dup”, consisting of those unique names where there are confirmed multiple instances. As each last name is checked by the program it first checks to see if the name is a new unique name. 

If the name is not in ‘dup’ but is in ‘first_time_seen’  then we have just discovered a new duplicate. Therefore the program appends the name to ‘dup’. If the name being checked is neither in ‘dup‘ nor in first_time_seen, then it is an entirely new name (so far). It then is appended to ‘first_time_seen’.


NOTE:  In a subsequent discussion with Christian, he described simpler ways of checking for instances of the same last name: 
a. ‘reverse’, ‘string.split’   b. ‘new value counts’ 
However, my way of doing it was a good exercise for me in working out the underlying logic of for loops and elif. len(dup) would also give the count. The code and output below shows that there are indeed passengers who share a last name.


In [78]:
print('Passengers who shared a last name')
names=df['Name']
dup=[]
first_time_seen=[]
x=0

for each in names:
    last_name=each.rsplit(maxsplit=1)
    ls=last_name[1]
    ls=ls+' ' # Adding a whitespace to 'ls' to ensure e.g. that Carr is not confused with Carrou
# Check if it is in 'dup' already
    if ls in dup:
        #  do nothing
        x=x #  x=x is a work-around because cannot have an empty line
    elif (ls  in first_time_seen):
        dup.append(ls) # Adding a comma between last names in list
        x=x+1 # Each added last name counts as 1 more
    else:
        first_time_seen.append(ls) 

print('There are', x, 'duplicates:', dup) # list of shared last names
print('_____________________________________________________________________')

Passengers who shared a last name
There are 133 duplicates: ['Palsson ', 'Planke ', 'Andersson ', 'Goodwin ', 'Fortune ', 'Moran ', 'Turpin ', 'Nasser ', 'White ', 'Nicola-Yarred ', 'Futrelle ', 'Ford ', 'Williams ', 'Panula ', 'Skoog ', 'Rice ', 'Johnson ', 'Sage ', 'Asplund ', 'Navratil ', 'Olsen ', 'Backstrom ', 'Harris ', 'Lefebre ', 'Zabour ', 'Attalah ', 'Strom ', 'Smith ', 'Baxter ', 'Allison ', 'Kantor ', 'Caldwell ', 'Goldsmith ', 'McCoy ', 'Graham ', 'Pears ', 'Brown ', 'Arnold-Franchi ', "O'Brien ", 'Meyer ', 'Gustafsson ', 'Holverson ', 'Carter ', 'Newell ', 'Sandstrom ', 'Johansson ', 'Olsson ', 'Jussila ', 'Hakkarainen ', 'Hart ', 'Minahan ', 'Rosblom ', 'Richards ', 'Mellinger ', 'West ', 'Baclini ', 'Braund ', 'Bishop ', 'Hoyt ', 'Coutts ', 'Hagland ', 'Calic ', 'Castellana ', 'Webber ', 'Hippach ', 'Quick ', 'Elias ', 'Peter ', 'Cacic ', 'Beane ', 'Davies ', 'Taussig ', 'Flynn ', 'Kelly ', 'Silvey ', 'Thayer ', 'Bourke ', 'Impe ', 'Boulos ', 'Gordon ', 'Jacobsohn ', 'L

5) pivot-tabels   
    

In [79]:
print('How many passengers travelled, respectively, first, second and third class') 

pivot3=pd.pivot_table(df, index=['Pclass'], values=['Name'], aggfunc={'count'})
print(pivot3)
print('_____________________________________________________________________')

How many passengers travelled, respectively, first, second and third class
        Name
       count
Pclass      
1        216
2        184
3        487
_____________________________________________________________________


In [80]:
print('The passenger class with the most dead')

pivot4=pd.pivot_table(df, index=['Survived', 'Pclass'], values=['Name'], aggfunc={'count'})
print(pivot4)
print('_____________________________________________________________________')
print('END')

The passenger class with the most dead
                 Name
                count
Survived Pclass      
0        1         80
         2         97
         3        368
1        1        136
         2         87
         3        119
_____________________________________________________________________
END


Third class passengers had the most dead, 368.


### CONCLUSION

It is possible to download a .csv file and use Python and pandas to create dataframes, describe the data set and create descriptive statistics. This can be used to answer research questions about the Titanic's passengers, such as the correlation between age and survival, or whether different passenger classes had different survival rates.


### BIBLIOGRAPHY
Anderson, R. E. (1992) ‘ACM Code of Ethics and Professional Conduct’, Communications of the ACM, 35(5), pp. 94–99. doi: 10.1145/129875.129885.

Association for Computing Machinery (2018) ‘Artifact Review and Badging’, p. 5. Available at: https://www.acm.org/publications/policies/artifact-review-badging.

Downey, A. (2015) Think Python: How to Think Like a Computer Scientist. 2nd, version edn. Needham, MA: Green Tea Press. Available at: http://www.thinkpython2.com.

Ignatow, G. and Milhacea (2018) Introduction to Text Mining. Menlo Park, CA: SAGE Publications, Inc.
Lund, H. (2019) ‘OPD uge 36 SLIDES Modul_1_introduktion_Open Data Science’. Copenhagen: University of Copenhagen, Institute for information Studies, pp. 1–29.

Thompson, S. A. and Warzel, C. (2019) ‘Twelve Million Phones, One Dataset, Zero Privacy’, The New York Times, 19 December. Available at: https://www.nytimes.com/interactive/2019/12/19/opinion/location-tracking-cell-phone.html.

Vicente-Saez, R. and Martinez-Fuentes, C. (2018) ‘Open Science now: A systematic literature review for an integrated definition’, Journal of Business Research, 88(July), pp. 428–436. doi: 10.1016/j.jbusres.2017.12.043.

Wilkinson, M. D. et al. (2016) ‘Comment: The FAIR Guiding Principles for scientific data management and stewardship’, Scientific Data, 3, pp. 1–9. doi: 10.1038/sdata.2016.18.
